File: lib/sync.ts
import { fetchHouseMembers, fetchSenateMembers, getStateCode } from "./congress"
import {
  fetchAllCongressTrades,
  fetchBulkCongressTrades,
  tradeToDbRow,
  type RawCongressTrade,
} from "./quiver"
import { fetchFecTotals, fetchPacDonations } from "./fec"
import { resolveFecCandidate, currentCycle, aggregateDonors } from "./profile"
import { refreshAllRankings, mapWithConcurrency } from "./rankings"
import {
  upsertMembers,
  upsertTrades,
  upsertFecCandidate,
  replacePacDonations,
  type DbMember,
  type DbTrade,
  type DbPacDonation,
} from "./db"

function resolveCongressApiKey(): string {
  const apiKey = process.env.CONGRESS_API_KEY ?? process.env.CONGRESS_GOV_API_KEY
  if (!apiKey) throw new Error("Missing CONGRESS_API_KEY")
  return apiKey
}

function lastName(listName: string): string {
  return listName.split(",")[0]?.trim() || listName
}

const NAME_SUFFIXES = new Set(["jr", "sr", "ii", "iii", "iv", "v"])

// Order-independent, accent-insensitive name key so the roster's "Last, First"
// and Quiver's "First Last" collapse to the same value (e.g. "Van Hollen, Chris"
// and "Chris Van Hollen" both -> "chris hollen van").
function nameTokenKey(name: string): string {
  return name
    .normalize("NFD")
    .replace(/[^\x00-\x7F]/g, "")
    .toLowerCase()
    .split(/[^a-z]+/)
    .filter((token) => token && !NAME_SUFFIXES.has(token))
    .sort()
    .join(" ")
}

function buildBioguideByName(
  members: { id: string; name: string }[]
): Map<string, string> {
  const byKey = new Map<string, string>()
  const ambiguous = new Set<string>()
  for (const member of members) {
    const key = nameTokenKey(member.name)
    if (!key) continue
    const existing = byKey.get(key)
    if (existing && existing !== member.id) ambiguous.add(key)
    else byKey.set(key, member.id)
  }
  // Drop names shared by two members rather than risk attributing a trade to the
  // wrong person; those stay unresolved (and are dropped on insert, as before).
  for (const key of ambiguous) byKey.delete(key)
  return byKey
}

function attachMissingBioguides(
  trades: RawCongressTrade[],
  byName: Map<string, string>
): number {
  let resolved = 0
  for (const trade of trades) {
    if (trade.Bioguide || !trade.Representative) continue
    const id = byName.get(nameTokenKey(trade.Representative))
    if (id) {
      trade.Bioguide = id
      resolved += 1
    }
  }
  return resolved
}

// Quiver's congress feed only carries a BioGuideID for House members, so every
// Senate disclosure arrives without one and tradeToDbRow drops it — which is why
// senator profiles showed no trades. Backfill the id from the synced roster by
// name so Senate trades persist too. Degrades to a no-op if the roster can't be
// loaded, leaving the prior (House-only) behavior untouched.
async function resolveMissingBioguides(trades: RawCongressTrade[]): Promise<number> {
  if (!trades.some((trade) => !trade.Bioguide && trade.Representative)) return 0
  try {
    const apiKey = resolveCongressApiKey()
    const [house, senate] = await Promise.all([
      fetchHouseMembers(apiKey),
      fetchSenateMembers(apiKey),
    ])
    return attachMissingBioguides(trades, buildBioguideByName([...house, ...senate]))
  } catch (error) {
    console.warn("[sync] bioguide name-resolution skipped:", error)
    return 0
  }
}

// Snapshot the current House + Senate rosters into `members`.
export async function syncMembers(): Promise<{ house: number; senate: number }> {
  const apiKey = resolveCongressApiKey()
  const [house, senate] = await Promise.all([
    fetchHouseMembers(apiKey),
    fetchSenateMembers(apiKey),
  ])

  const rows: DbMember[] = [
    ...house.map((m) => ({
      bioguide_id: m.id,
      name: m.name,
      party: m.party,
      state: m.state,
      district: m.district,
      chamber: "house" as const,
      image_url: m.imageUrl ?? null,
    })),
    ...senate.map((m) => ({
      bioguide_id: m.id,
      name: m.name,
      party: m.party,
      state: m.state,
      district: null,
      chamber: "senate" as const,
      image_url: m.imageUrl ?? null,
    })),
  ]

  await upsertMembers(rows)
  return { house: house.length, senate: senate.length }
}

export async function syncTrades(): Promise<{ fetched: number; persisted: number }> {
  const apiKey = process.env.QUIVER_API_KEY
  if (!apiKey) throw new Error("Missing QUIVER_API_KEY")

  const trades = await fetchAllCongressTrades(apiKey, { forceRefresh: true })
  const resolved = await resolveMissingBioguides(trades)
  const rows = trades
    .map(tradeToDbRow)
    .filter((row): row is DbTrade => row !== null)

  await upsertTrades(rows)
  console.log(`[sync] syncTrades: resolved ${resolved} trade(s) to a bioguide by name`)
  return { fetched: trades.length, persisted: rows.length }
}

// Full multi-year history (Quiver's bulk endpoint, ~2020→present). Upsert-only
// and idempotent — unified trade ids mean these rows merge with the 15-minute
// live sync instead of duplicating. Run on a slow cadence (daily); the live
// `syncTrades` keeps the most-recent disclosures fresh between runs.
export async function backfillTrades(): Promise<{
  fetched: number
  persisted: number
}> {
  const apiKey = process.env.QUIVER_API_KEY
  if (!apiKey) throw new Error("Missing QUIVER_API_KEY")

  const trades = await fetchBulkCongressTrades(apiKey)
  const resolved = await resolveMissingBioguides(trades)
  const rows = trades
    .map(tradeToDbRow)
    .filter((row): row is DbTrade => row !== null)

  await upsertTrades(rows)
  console.log(`[sync] backfillTrades: resolved ${resolved} trade(s) to a bioguide by name`)
  return { fetched: trades.length, persisted: rows.length }
}

export async function syncRankings(): Promise<{ house: number; senate: number }> {
  const apiKey = resolveCongressApiKey()
  const { house, senate } = await refreshAllRankings(apiKey)
  return { house: house.byNetWorth.length, senate: senate.byNetWorth.length }
}
const FEC_CONCURRENCY = 4
const FEC_DELAY_MS = 150
export async function syncFec(): Promise<{ members: number; resolved: number }> {
  const apiKey = resolveCongressApiKey()
  const fecKey = process.env.FEC_API_KEY
  if (!fecKey) throw new Error("Missing FEC_API_KEY")

  const [house, senate] = await Promise.all([
    fetchHouseMembers(apiKey),
    fetchSenateMembers(apiKey),
  ])

  const targets = [
    ...house.map((m) => ({
      id: m.id,
      name: lastName(m.name),
      state: m.state,
      office: "H" as const,
      preferIncumbent: false,
      perPage: 10,
    })),
    ...senate.map((m) => ({
      id: m.id,
      name: lastName(m.name),
      state: m.state,
      office: "S" as const,
      preferIncumbent: true,
      perPage: 20,
    })),
  ]

  const cycle = currentCycle()
  let resolved = 0

  await mapWithConcurrency(
    targets,
    FEC_CONCURRENCY,
    async (target) => {
      const ref = await resolveFecCandidate({
        name: target.name,
        stateCode: getStateCode(target.state),
        office: target.office,
        preferIncumbent: target.preferIncumbent,
        perPage: target.perPage,
      })
      if (!ref) return
      resolved += 1

      const totals = await fetchFecTotals(ref.candidateId, fecKey)
      await upsertFecCandidate({
        bioguide_id: target.id,
        candidate_id: ref.candidateId,
        committee_ids: ref.committeeIds,
        total_raised: totals?.receipts ?? 0,
        total_spent: totals?.disbursements ?? 0,
        cycle,
      })

      const { allDonations } = await fetchPacDonations(ref.committeeIds, fecKey)
      const rows: DbPacDonation[] = aggregateDonors(allDonations).map((d) => ({
        bioguide_id: target.id,
        pac_name: d.pacName,
        amount: d.amount,
        cycle,
      }))
      await replacePacDonations(target.id, rows)
    },
    FEC_DELAY_MS
  )

  return { members: targets.length, resolved }
}

File: backend/app/services/sync.py
"""Scheduled ETL jobs (port of lib/sync.ts)."""

import asyncio
import logging

from ..clients.congress import fetch_house_members, fetch_senate_members, get_state_code
from ..clients.fec import fetch_fec_totals, fetch_pac_donations
from ..clients.quiver import (
    fetch_all_congress_trades,
    fetch_bulk_congress_trades,
    trade_to_db_row,
)
from ..config import fec_api_key, quiver_api_key, require_congress_api_key
from ..core.db import (
    replace_pac_donations,
    upsert_fec_candidate,
    upsert_members,
    upsert_trades,
)
from ..core.util import build_bioguide_by_name, map_with_concurrency, name_token_key
from .disclosures import refresh_disclosure_net_worth
from .profile import aggregate_donors, current_cycle, resolve_fec_candidate
from .rankings import refresh_all_rankings
from .entity_cards import refresh_entity_cards
from .stock_leaderboard import refresh_stock_performance
from .trade_features import refresh_trade_features

logger = logging.getLogger("sync")


def _last_name(list_name: str) -> str:
    return list_name.split(",")[0].strip() or list_name


def _attach_missing_bioguides(trades: list[dict], by_name: dict[str, str]) -> int:
    resolved = 0
    for trade in trades:
        if trade.get("Bioguide") or not trade.get("Representative"):
            continue
        bioguide_id = by_name.get(name_token_key(trade["Representative"]))
        if bioguide_id:
            trade["Bioguide"] = bioguide_id
            resolved += 1
    return resolved


async def _resolve_missing_bioguides(trades: list[dict]) -> int:
    """Quiver's congress feed only carries a BioGuideID for House members, so
    every Senate disclosure arrives without one and trade_to_db_row drops it.
    Backfill the id from the synced roster by name so Senate trades persist too.
    Degrades to a no-op if the roster can't be loaded."""
    if not any(not t.get("Bioguide") and t.get("Representative") for t in trades):
        return 0
    try:
        api_key = require_congress_api_key()
        house, senate = await asyncio.gather(
            fetch_house_members(api_key), fetch_senate_members(api_key)
        )
        return _attach_missing_bioguides(
            trades, build_bioguide_by_name([*house, *senate])
        )
    except Exception as error:
        logger.warning("bioguide name-resolution skipped: %s", error)
        return 0


async def sync_members() -> dict:
    """Snapshot the current House + Senate rosters into `members`."""
    api_key = require_congress_api_key()
    house, senate = await asyncio.gather(
        fetch_house_members(api_key), fetch_senate_members(api_key)
    )

    rows = [
        *(
            {
                "bioguide_id": m["id"],
                "name": m["name"],
                "party": m["party"],
                "state": m["state"],
                "district": m["district"],
                "chamber": "house",
                "image_url": m.get("imageUrl"),
            }
            for m in house
        ),
        *(
            {
                "bioguide_id": m["id"],
                "name": m["name"],
                "party": m["party"],
                "state": m["state"],
                "district": None,
                "chamber": "senate",
                "image_url": m.get("imageUrl"),
            }
            for m in senate
        ),
    ]

    await upsert_members(rows)
    return {"house": len(house), "senate": len(senate)}


async def sync_trades() -> dict:
    api_key = quiver_api_key()
    if not api_key:
        raise RuntimeError("Missing QUIVER_API_KEY")

    trades = await fetch_all_congress_trades(api_key, force_refresh=True)
    resolved = await _resolve_missing_bioguides(trades)
    rows = [row for row in (trade_to_db_row(t) for t in trades) if row is not None]

    await upsert_trades(rows)
    logger.info("sync_trades: resolved %s trade(s) to a bioguide by name", resolved)
    return {"fetched": len(trades), "persisted": len(rows)}


async def backfill_trades() -> dict:
    api_key = quiver_api_key()
    if not api_key:
        raise RuntimeError("Missing QUIVER_API_KEY")

    trades = await fetch_bulk_congress_trades(api_key)
    resolved = await _resolve_missing_bioguides(trades)
    rows = [row for row in (trade_to_db_row(t) for t in trades) if row is not None]

    await upsert_trades(rows)
    logger.info("backfill_trades: resolved %s trade(s) to a bioguide by name", resolved)
    return {"fetched": len(trades), "persisted": len(rows)}


async def sync_rankings() -> dict:
    api_key = require_congress_api_key()
    result = await refresh_all_rankings(api_key)
    return {
        "house": len(result["house"]["byNetWorth"]),
        "senate": len(result["senate"]["byNetWorth"]),
    }


async def sync_stock_performance() -> dict:
    """Best-performing-stocks leaderboard (Congress P/L on disclosed buys).
    Reads the persisted trades + price history, so it has no Quiver fan-out of
    its own and runs on a slow cadence. The holdings leaderboard has no job
    here — it is filled as a side effect of refresh_all_rankings."""
    rows = await refresh_stock_performance()
    return {"stocks": len(rows)}


async def sync_trade_features() -> dict:
    """Rebuild the RAG feature layer (trade_features + asset_class_stats) from
    the persisted trades, then refresh the semantic entity cards built on top
    of it. Reads the DB plus price history for the priced universe, so it has
    no Quiver fan-out and runs daily."""
    features = await refresh_trade_features()
    cards = await refresh_entity_cards()
    return {**features, **cards}


async def sync_disclosures() -> dict:
    """Rebuild the annual financial-disclosure net-worth estimates that fill
    the rankings' net-worth gaps for members Quiver has no live figure for.
    Downloads/parses ~500 filings, so it runs on a weekly cadence — annual
    filings barely change between runs."""
    estimates = await refresh_disclosure_net_worth()
    return {"members": len(estimates)}


FEC_CONCURRENCY = 4
FEC_DELAY_MS = 150


async def sync_fec() -> dict:
    api_key = require_congress_api_key()
    fec_key = fec_api_key()
    if not fec_key:
        raise RuntimeError("Missing FEC_API_KEY")

    house, senate = await asyncio.gather(
        fetch_house_members(api_key), fetch_senate_members(api_key)
    )

    targets = [
        *(
            {
                "id": m["id"],
                "name": _last_name(m["name"]),
                "state": m["state"],
                "office": "H",
                "preferIncumbent": False,
                "perPage": 10,
            }
            for m in house
        ),
        *(
            {
                "id": m["id"],
                "name": _last_name(m["name"]),
                "state": m["state"],
                "office": "S",
                "preferIncumbent": True,
                "perPage": 20,
            }
            for m in senate
        ),
    ]

    cycle = current_cycle()
    resolved = 0

    async def process(target: dict) -> None:
        nonlocal resolved
        ref = await resolve_fec_candidate(
            name=target["name"],
            state_code=get_state_code(target["state"]),
            office=target["office"],
            prefer_incumbent=target["preferIncumbent"],
            per_page=target["perPage"],
        )
        if not ref:
            return
        resolved += 1

        totals = await fetch_fec_totals(ref["candidateId"], fec_key)
        await upsert_fec_candidate(
            {
                "bioguide_id": target["id"],
                "candidate_id": ref["candidateId"],
                "committee_ids": ref["committeeIds"],
                "total_raised": (totals or {}).get("receipts") or 0,
                "total_spent": (totals or {}).get("disbursements") or 0,
                "cycle": cycle,
            }
        )

        donations = await fetch_pac_donations(ref["committeeIds"], fec_key)
        rows = [
            {
                "bioguide_id": target["id"],
                "pac_name": d["pacName"],
                "amount": d["amount"],
                "cycle": cycle,
            }
            for d in aggregate_donors(donations["allDonations"])
        ]
        await replace_pac_donations(target["id"], rows)

    await map_with_concurrency(targets, FEC_CONCURRENCY, process, FEC_DELAY_MS)

    return {"members": len(targets), "resolved": resolved}
